### ISSUE TITLE
Required layer inference is incorrect when requiring two layers of the same outer type.

### ISSUE THREAD

**Issue by JonNorman** [zio/zio#7677]
Hi there 👋 

This is the first issue I've raised here, please let me know if I've missed anything.

**TL;DR**: In attempting to use the `ZLayer` pattern to construct my services (as described in a number of places e.g. [here](https://softwaremill.com/structuring-zio-2-applications/)), if a service requires two layers with the same outer type but different inner types — e.g. `Seq[Int]` and `Seq[String]` — then the compiler errors with a message asking for a union of those types — `Seq[Int & String]` — regardless of whether two layers of `Seq[Int]` and `Seq[String]` are provided or not.

This is using `ZIO 2.0.1` and `Scala 3.2.0`

**Minimal reproduction**

I tried to reproduce it all as one file but encountered macro errors for the `provides` and so had to separate into `ExampleService` and the `Main` app.

```scala
// example/ExampleService.scala
package example

import zio.*

case class ExampleService(data1: Seq[String], data2: Seq[Int]):

  def start = ZIO.attempt(println(s"Data $data1 & $data2"))

object ExampleService:
  val layer = ZLayer.fromFunction(ExampleService.apply)
```

and

```scala
// example/Main.scala
package example

import zio.*

object Main extends ZIOAppDefault:

  override val run =
    ZIO
      .serviceWithZIO[ExampleService](_.start)
      .provide(
        ZLayer.succeed(Seq(1, 2, 3)),
        ZLayer.succeed(Seq("a", "b", "c")),
        ExampleService.layer
      )

```

**Expected**
Compiles correctly as all environment requirements are provided as layers (a `Seq[Int]` and a `Seq[String]` as required by the `ExampleService` layer).

**Actual**
Compilation error indicating that a union type of `Seq[Int & String]` must be provided:

```sh
ExampleService.layer -> List(scala.collection.immutable.Seq[scala.Predef.String & scala.Int])
[error] -- Error: ~/src/main/scala/example/Main.scala:10:14 
[error]  8 |    ZIO
[error]  9 |      .serviceWithZIO[ExampleService](_.start)
[error] 10 |      .provide(
[error]    |    ^
[error]    |
[error]    |
[error]    |  ──── ZLAYER ERROR ────────────────────────────────────────────────────
[error]    |
[error]    |   Please provide a layer for the following type:
[error]    |
[error]    |     Required by ExampleService.layer
[error]    |     1. scala.collection.immutable.Seq[scala.Predef.String & scala.Int]
[error]    |        
[error]    |  ──────────────────────────────────────────────────────────────────────
[error]    |
[error] 11 |        ZLayer.succeed(Seq(1, 2, 3)),
[error] 12 |        ZLayer.succeed(Seq("a", "b", "c")),
[error] 13 |        ExampleService.layer
[error] 14 |      )
[error] one error found
[error] (Compile / compileIncremental) Compilation failed
[error] Total time: 3 s, completed 6 Jan 2023, 16:13:28
```

**Comment by adamgfraser** [comment]
Thanks for reporting!

An easy workaround for this is to create a new type for each of your services, for example `final case class Letters(letters: Chunk[String])` and `final case class Numbers(numbers: Chunk[Int])`.

**Comment by JonNorman** [comment]
Thanks @adamgfraser! That does indeed work, we'll go with that for now 👍 

Appreciate the rapid response and solution!

**Comment by davidlar** [comment]
I get the same problem after trying to switch to scala 3. In my case it's not collections like Seq, but instead our own classes with type parameters. We use them e.g. when we have multiple layers with the same implementation.

**Comment by adamgfraser** [comment]
Looking at this more the compiler is correct here. The unification of `Seq[String]` and `Seq[Int]` is `Seq[String & Int]` due to covariance. To specify that you want two different collections here create a new nominal type as described above.

**Comment by JonNorman** [comment]
@adamgfraser I appreciate that this makes sense from a strict types perspective as `Seq[String & Int]` satisfies both individual requirements of `Seq[String]` and `Seq[Int]`. This works for the consumers of the dependencies i.e. the services.

However, in my opinion it contravenes the expectations of dependency-injection, especially as the _provider_ of the dependencies. For example, what would it even mean to provide a `Seq[String & Int]` and is it not fairly arbitrary that, because they have the same outer-type (but are not the same exact type), that the dependencies must be combined into a single dependency?

Perhaps I have become too used to the simplicity and pragmatism of ZIO that I find even simple workarounds slightly objectionable if they seem to contravene this ethos 😀. With that in mind, I'd also like to say thank you for all your efforts and constant attention that drives projects like these - it is much, much appreciated.

**Comment by adamgfraser** [comment]
@JonNorman I totally hear you but we also have to work with Scala's type system. Consider this simple example:

```scala
trait Box[+A]

trait Animal
trait Cat extends Animal
trait Dog extends Animal
  
val a = ZIO.environment[Box[Cat]]
val b = ZIO.environment[Box[Dog]]

val c: ZIO[Box[Cat & Dog], Nothing, Unit]  =
  a *> b *> ZIO.unit
``` 

I agree that `Box[Cat & Dog]` is probably not what you want here but as you can see it has nothing to do with the way layer construction works and nothing to do even with ZIO other than the fact that environment type is contravariant. I could create a "toy ZIO" example that does exactly the same thing.

Note that if `Box` was invariant you would get your desired type of `Box[Cat] & Box[Dog]` there.

This is basically a consequence of deleting `Has` and making the environment type covariant which overall has been extremely positive but one thing that `Has` did was create an invariant "wrapper" around each type.

So while I agree with you I don't think we can work around Scala's type system and the fact that when you compose these two workflows you get a workflow that requires a `Box[Cat & Dog]`. The easy way to solve this problem is to create a new type like `final case class Dogs(dogs: Box[Dog])`.

Thanks for the kind words and thanks for using ZIO and raising this!
